extensions/2.0/Khronos/KHR_lights_environment/README.md

```diff
@@ -0,0 +1,143 @@
+
+# KHR_lights_environment
+
+## Contributors
+
+* Norbert Nopper, <mailto:nopper@ux3d.io>
+* Rickard Sahlin, <mailto:rickard.sahlin@ikea.com>
+* Gary Hsu, Microsoft, <mailto:garyhsu@microsoft.com>
+* Mike Bond, Adobe, <mailto:mbond@adobe.com>
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+This extension may use KHR_texture_basisu, KHR_texture_float_bc6h or KHR_texture_float_rgb9e5  
+
+## Overview
+This extension provides the ability to define image-based lights in a glTF scene.  
+Image-based lights consist of an environment map that represents specular radiance for the scene as well as irradiance information.  
+This can be used on it's own - ie a glTF asset with only environment map data - for a usecase where the IBL needs to be distributed.  
+It can also be used together with model data, for usecases where a model shall be displayed in a defined environment.  
+
+Many 3D tools and engines support image-based global illumination but the exact technique and data formats employed vary.  
+Using this extension, tools can export and engines can import image-based lights and the result should be highly consistent.   
+
+This extension specifies exactly one way to format and reference the environment map to be used.  
+The goals of this are two-fold.  
+First, it makes implementing support for this extension easier.  
+Secondly, it ensures that rendering of the image-based lighting is consistent across runtimes.
+
+A conforming implementation of this extension must be able to load the image-based environment data and render the PBR materials using this lighting.  
+
+This extension is based on EXT_lights_image_based, with added support for KTX2 as a carrier for [cubemap and] HDR texture formats.  
+
+The environment light is defined by a cubemap.  
+[TBD - shall a separate KHR_texture_cubemap extension be created?]
+Cubemaps can be supplied pre-filtered, by including the pre-filtered mip-levels, otherwise filtering will be performed at load time.  
+[TBD - point to relevant filtering algorithm]  
+If a compressed texture format is used then pre-filtered mip-levels must be included.  
+
+## Declaring an environment light
+
+The KHR_lights_environment extension defines an array of image-based lights at the root of the glTF and then each scene can reference one.  
+Each environment light definition consists of a single cubemap that describes the specular radiance of the scene, the l=2 spherical harmonics coefficients for diffuse irradiance and intensity values.  
+The cubemap is defined by texture references to a KTX2 file containing a cubemap, these files can contain compressed textures using KHR_texture_basisu  
+or use a float texture format as defined by KHR_texture_float_bc6h or KHR_texture_float_rgb95e.   
+
+
+```json
+"extensions": {
+    "KHR_lights_environment" : {
+        "lights": [
+            {
+                "intensity": 1.0,
+                "irradianceCoefficients": [...3 x 9 array of floats...],
```

Comment:
This is in my opinion derived data from the specular cube map.  Thus I question whether we need to have it here.  It seems like pre-mature optimization and also increases implementation work because now all engines need to support Spherical Harmonics.

Reply:
I would like to have both the Lambertian contribution from the spherical harmonics and the specular from the texture.
As a runtime format I think it makes sense to bake this data offline since it is so small - otherwise implementations needs to do the convolution in rumtime (which I think is more cumbersome than implementing the spherical harmonics)

For the implementation, I think that is quite straightforward and I'll make sure to include links to examples.

Reply:
> As a runtime format I think it makes sense to bake this data offline since it is so small

Three.js does not use the SH do add the Lambertian contribution in the default workflow.  It just derived it from the convolution of the specular maps.

There are significant limitations to SH representations, even though I used to be a supporter -- and I did write the initial implementation that turned into Three.js SH support: https://github.com/mrdoob/three.js/pull/6405/files

For example UE4 no longer supports SH for sly lights (our specific use case) and instead just uses cube maps.  They say: "Prior to Unreal Engine 4.18, Static Sky Lights used to be represented by Lightmass with a 3rd Order Spherical Harmonic which didn't capture the detail that could be present in a sunrise or sunset."  https://docs.unrealengine.com/en-US/BuildingWorlds/LightingAndShadows/LightTypes/SkyLight/index.html

The changing of the state-of-the-art implementation leads me to want to avoid presupposing unnecessarily the implementation, rather than a description of the physical inputs (which are much less likely to become obsolete.)

Reply:
This is why I would like to have the option of Spherical Harmonics - choose solution based on your usecase.

If diffuse light does not encode well in the SH - skip it and let the implementation revert to fragment shader lookup.
If your diffuse light does encode well in the SH - then you free up a uniform sampler slot and get the diffuse light contribution in the vertex stage.

Reply:
Not an expert by any means so I don't have very strong opinions on if SH are a reasonable enough approximation, but if they are it does seem in line with the goals of GLTF to have these pre computed, as the goal is to have as direct a path from disk to gpu memory as possible.

Reply:
I agree @netpro2k, SH are approximation - they will not be able to realistically capture all environments.
For low frequency type of environments they do a good job, for high frequency or HDR type of environments they are not ideal.
I will keep them based on the glTF goal of providing data that reduces load time processing.